IP address filtering

kuenishi opened the pull request:
define policy record and add bucket policy set/get/delete/check functions
basic design is just an extension of ACL, which is stored as a metadata of moss.bucket/<bucketname> riakc_object.

kuenishi committed:
define policy record and add bucket policy set/get/delete/check functions
and implement check logic (issue #32)

kellymclaughlin commented on include/stanchion.hrl:
```diff
@@ -45,10 +46,122 @@
 -define(MOSS_USER, #rcs_user_v2).
 -define(RCS_USER, #rcs_user_v2).
 -define(MD_ACL, <<"X-Moss-Acl">>).
+-define(MD_POLICY, <<"X-Moss-Policy">>).
```
Let's avoid using `Moss` in anything new. Some of the uses we cannot get rid of, but let's not add anymore.

kellymclaughlin commented on include/stanchion.hrl:
```diff
@@ -45,10 +46,122 @@
 -type acl() :: acl1() | acl2().
 
+
+% S3 Actions
```
I think it would be good to try and modularize and decouple S3 specific things where possible. In that vein can we move the S3 specific stuff to a separate header file?

kuenishi replied:
OK, I'll try and think how policy API should be - or can be independent of S3.

kellymclaughlin commented on src/stanchion_policy.erl:
```diff
@@ -0,0 +1,676 @@
+
+-include("stanchion.hrl").
+-include_lib("webmachine/include/wm_reqdata.hrl").
+-include_lib("eunit/include/eunit.hrl").
```
Might want to `ifdef(TEST)` around this prior to a merge.

kuenishi committed:
separate s3-dependent codes, most of them is moved to riak_cs.
stanchion does not parse JSON, but just stores it into riak
which was verified at riak_cs process. riak_cs parse and checks
policy on every request from client.

kellymclaughlin commented on src/stanchion_utils.erl:
```diff
@@ -261,14 +265,48 @@ pow(Base, Power, Acc) ->
     UpdBucketObj = riakc_obj:update_metadata(UpdBucketObj0, MetaData),
     riakc_pb_socket:put(RiakPid, UpdBucketObj).
 
+make_new_metadata(MetaVals, {acl, Acl})->
```
This looks odd to me. If there is any other information in the `?MD_USERMETA` dictionary that is passed in it will be lost in this function call. The only things preserved are the `?MD_ACL` and `?MD_POLICY` entries. It's also destructive of any other entries in the toplevel metadata dictionary for the riak_object. Seems like something like this would be better:

```
put_bucket(BucketObj, OwnerId, AclOrPolicy, RiakPid) ->    
    UpdBucketObj0 = riakc_obj:update_value(BucketObj, OwnerId),
    MD = case riakc_obj:get_metadatas(UpdBucketObj0) of
             [] -> % create                                                                                                                                                          
                 {acl, Acl} = AclOrPolicy,
                 M0 = [{?MD_ACL, term_to_binary(Acl)}],
                 dict:from_list([{?MD_USERMETA, M0}]);            
            [MD0] -> MD0;             _E ->
                 MsgData = {siblings, riakc_obj:key(BucketObj)},
                 _ = lager:error("bucket has siblings: ~p", [MsgData]),
                 throw(MsgData) % @TODO: data broken; handle this                                                                                                                    
           end,
    MetaVals = dict:fetch(?MD_USERMETA, MD),
    UserMetaData = make_new_user_metadata(MetaVals, AclOrPolicy),
    MetaData = make_new_metadata(MD, UserMetaData),
    UpdBucketObj = riakc_obj:update_metadata(UpdBucketObj0, MetaData),
    riakc_pb_socket:put(RiakPid, UpdBucketObj).

make_new_metadata(MD, UserMeta) ->
    [{?MD_USERMETA, UserMeta} | proplists:delete(?MD_USERMETA, MD)].

make_new_user_metadata(MetaVals, {acl, Acl})->    
    [{?MD_ACL, term_to_binary(Acl)} |     proplists:delete(?MD_ACL, MetaVals)];
make_new_user_metadata(MetaVals, {policy, Policy}) ->
    [{?MD_POLICY, term_to_binary(Policy)} |
     proplists:delete(?MD_POLICY, MetaVals)];
make_new_user_metadata(MetaVals, delete_policy) ->
    proplists:delete(?MD_POLICY, MetaVals).
```

That way attributes in the toplevel metadata structure are preserved as well as other entries in the user metdata structure.

kuenishi replied:
I thought there are no other metadata other than ?MD_USERMETA ( is this true? ) - anyway it might cause bugs in case of other metadata added in future. I'll fix this after next Monday.

kuenishi committed:
fix bug that updating acl or policy won't preserve other metadatas
stored in Riak

kellymclaughlin commented:
+1 to merge
